engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -779,101 +944,285 @@ static void CheckTextureResource(lua_State* L, int i, const char* field_name, dm
+    DestroyTextureImage(texture_image, create_params.m_Buffer == 0);
+
+    if (res != dmResource::RESULT_OK)
     {
-        return luaL_error(L, "Unable to set texture, unsupported texture format '%s'.", dmGraphics::GetTextureFormatLiteral(format));
+        return ReportPathError(L, res, create_params.m_PathHash);
     }
 
-    // TODO: To support this, we need to supply separate buffers for each side as an option, or offsets into the buffer where each side is located
-    if ((tex_type == dmGraphics::TextureImage::TYPE_CUBEMAP || tex_type == dmGraphics::TextureImage::TYPE_2D_ARRAY) && compression_type != dmGraphics::TextureImage::COMPRESSION_TYPE_DEFAULT)
+    dmGameObject::AddDynamicResourceHash(create_params.m_Collection, create_params.m_PathHash);
+
+    if (g_ResourceModule.m_LoadRequests.Full())
     {
-        return luaL_error(L, "Compression type %d requested for texture %s with type '%s', but this is currently not supported.",
-            (int) compression_type, path_str, dmGraphics::GetTextureTypeLiteral(type));
+        g_ResourceModule.m_LoadRequests.Allocate(16);
     }
 
-    MakeTextureImage(width, height, max_mipmaps, tex_bpp, tex_type, tex_format, compression_type, buffer, &texture_image);
+    dmGraphics::TextureCreationParams texture_create_params;
+    texture_create_params.m_Type = create_params.m_Type;
 
-    dmArray<uint8_t> ddf_buffer;
-    dmDDF::Result ddf_result = dmDDF::SaveMessageToArray(&texture_image, dmGraphics::TextureImage::m_DDFDescriptor, ddf_buffer);
-    assert(ddf_result == dmDDF::RESULT_OK);
+    dmGraphics::HTexture texture_dst = dmGraphics::NewTexture(g_ResourceModule.m_GraphicsContext, texture_create_params);
 
-    void* resource = 0x0;
-    dmResource::Result res = dmResource::CreateResource(g_ResourceModule.m_Factory, path_str, ddf_buffer.Begin(), ddf_buffer.Size(), &resource);
+    SetTextureAsyncRequest* request = new SetTextureAsyncRequest();
+    HOpaqueHandle request_handle = g_ResourceModule.m_LoadRequests.Put(request);
+    request->m_LuaState          = L;
+    request->m_TextureResource   = (TextureResource*) resource;
+    request->m_Handle            = request_handle;
+    request->m_CallbackInfo      = callback_info;
+    request->m_Buffer            = dst_buffer;
+    request->m_Texture           = texture_dst;
+    request->m_UseUploadBuffer   = use_upload_buffer;
+    request->m_PathHash          = create_params.m_PathHash;
 
-    DestroyTextureImage(texture_image, buffer == 0);
+    dmGraphics::TextureParams texture_params;
+    texture_params.m_Width  = create_params.m_Width;
+    texture_params.m_Height = create_params.m_Height;
+    texture_params.m_Format = create_params.m_Format;
 
-    if (res != dmResource::RESULT_OK)
+    dmBuffer::GetBytes(request->m_Buffer, (void**)&texture_params.m_Data, &texture_params.m_DataSize);
+    dmGraphics::SetTextureAsync(texture_dst, texture_params);
```

Comment:
We sidestep the resource system here for the update mainly for simplicity, but should we consider being able to update a texture resource without explicit syncronization? the texture resource uploads data via the same SetTextureAsync function here but does an explicit sync. This causes the app to hang/stall until the upload is complete, which is the point of this function to avoid such behaviour..